public/app/features/alerting/unified/mocks/grafanaRulerApi.ts

```diff
@@ -61,33 +62,103 @@ export const grafanaRulerRule: RulerGrafanaRuleDTO = {
   },
 };
 
-export const grafanaRulerGroup: RulerRuleGroupDTO = {
+export const grafanaRulerGroup: RulerRuleGroupDTO<RulerGrafanaRuleDTO> = {
   name: grafanaRulerGroupName,
   interval: '1m',
   rules: [grafanaRulerRule],
 };
 
-export const grafanaRulerGroup2: RulerRuleGroupDTO = {
+export const grafanaRulerGroup2: RulerRuleGroupDTO<RulerGrafanaRuleDTO> = {
   name: grafanaRulerGroupName2,
   interval: '1m',
   rules: [grafanaRulerRule],
 };
 
-export const grafanaRulerEmptyGroup: RulerRuleGroupDTO = {
+export const grafanaRulerEmptyGroup: RulerRuleGroupDTO<RulerGrafanaRuleDTO> = {
   name: 'empty-group',
   interval: '1m',
   rules: [],
 };
 
-export const namespaceByUid: Record<string, { name: string; uid: string }> = {
-  [grafanaRulerNamespace.uid]: grafanaRulerNamespace,
-  [grafanaRulerNamespace2.uid]: grafanaRulerNamespace2,
-};
+// AKA Folder
+interface GrafanaNamespace {
+  name: string;
+  uid: string;
+}
 
-export const namespaces: Record<string, RulerRuleGroupDTO[]> = {
-  [grafanaRulerNamespace.uid]: [grafanaRulerGroup, grafanaRulerGroup2],
-  [grafanaRulerNamespace2.uid]: [grafanaRulerEmptyGroup],
-};
+export class RulerTestDb {
+  private namespaces = new Map<string, string>(); // UID -> Name
+  private groupsByNamespaceUid = new Map<string, RulerRuleGroupDTO[]>();
```

Comment:
This looks a bit complicated, but the way Grafana handles folders and namespaces in Ruler is quite tricky:
- Ruler response uses folder paths as keys in the response, but on the other hand we need to use folderUid for any ruler interaction (e.g. fetching single namespace or group)
This class tries to mimic this behaviour of the backend

Reply:
I think overall its clean to keep this logic encapsulated in a class and then use in the handlers 👌 Nice